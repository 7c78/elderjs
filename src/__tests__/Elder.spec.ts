process.cwd = () => 'test';

jest.mock('path', () => ({
  resolve: (...strings) => strings.join('/').replace('./', '').replace('//', '/').replace('/./', '/'),
  join: (...strings) => strings.join('/').replace('./', '').replace('//', '/').replace('/./', '/'),
}));

jest.mock('../utils/getConfig', () => () => ({
  debug: {
    automagic: true,
  },
  locations: {
    srcFolder: './src/',
    buildFolder: './__ELDER__/',
    svelte: {
      ssrComponents: './___ELDER___/compiled/',
      clientComponents: './public/dist/svelte/',
    },
  },
  server: {
    prefix: '/dev',
  },
  build: {
    shuffleRequests: false,
    numberOfWorkers: 4,
  },
  typescript: true,
  plugins: {
    'elder-plugin-upload-s3': {
      dataBucket: 'elderguide.com',
      htmlBucket: 'elderguide.com',
      deployId: '11111111',
    },
  },
  hooks: {
    disable: false,
  },
}));

jest.mock('../utils/getHashedSvelteComponents', () => () => ({
  File1: 'entryFile1',
  File2: 'entryFile2',
}));

jest.mock('../utils/validations', () => ({
  validatePlugin: (i) => i,
  validateHook: (i) => i,
  validateRoute: (i) => i,
}));

jest.mock('../utils/prepareRunHook', () => () => () => Promise.resolve());

describe('#Elder', () => {
  beforeEach(() => jest.resetModules());

  it('throws when plugin not found', async () => {
    jest.mock('fs-extra', () => ({
      existsSync: () => false,
    }));
    // eslint-disable-next-line global-require
    const { Elder } = require('../Elder');
    await expect(async () => {
      await new Elder({ context: 'build', worker: false });
    }).rejects.toThrow('Plugin elder-plugin-upload-s3 not found in plugins or node_modules folder.');
  });

  it('srcPlugin found', async () => {
    jest.mock('fs-extra', () => ({
      existsSync: () => true,
    }));
    jest.mock(
      'test/src/plugins/elder-plugin-upload-s3/index.js',
      () => ({
        hooks: [
          {
            hook: 'customizeHooks',
            name: 'test hook',
            description: 'just for testing',
            run: jest.fn(),
            $$meta: {
              type: 'hooks.js',
              addedBy: 'validations.spec.ts',
            },
          },
        ],
        routes: {},
        config: {},
        name: 'test',
        description: 'test',
        init: jest.fn(),
      }),
      {
        virtual: true,
      },
    );
    // eslint-disable-next-line global-require
    const { Elder } = require('../Elder');
    const elder = await new Elder({ context: 'server', worker: false });
    await elder.bootstrap();
    expect(elder).toMatchSnapshot();
  });
});
