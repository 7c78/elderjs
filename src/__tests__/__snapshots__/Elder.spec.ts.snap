// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Elder hookSrcFile not found 1`] = `
Elder {
  "allRequests": Array [],
  "bootstrapComplete": Promise {},
  "data": Object {},
  "errors": Array [],
  "helpers": Object {
    "inlineSvelteComponent": [Function],
    "permalinks": Object {
      "route-a": [Function],
      "test-a": [Function],
      "test-b": [Function],
    },
  },
  "hooks": Array [
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds external helpers to helpers object",
      "hook": "bootstrap",
      "name": "elderAddExternalHelpers",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Builds the shortcode parser, parses shortcodes from the html returned by the route's html and appends anything needed to the stacks.",
      "hook": "shortcodes",
      "name": "elderProcessShortcodes",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds <meta charset=\\"UTF-8\\" /> to the head.",
      "hook": "stacks",
      "name": "elderAddMetaCharsetToHead",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" /> to the head.",
      "hook": "stacks",
      "name": "elderAddMetaViewportToHead",
      "priority": 10,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Sets up the default polyfill for the intersection observer",
      "hook": "stacks",
      "name": "elderAddDefaultIntersectionObserver",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "AddsSystemJs to beforeHydrateStack also add preloading of systemjs to the headStack.",
      "hook": "stacks",
      "name": "elderAddSystemJs",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Creates an HTML string out of the Svelte layout and stacks.",
      "hook": "compileHtml",
      "name": "elderCompileHtml",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Log any errors to the console.",
      "hook": "error",
      "name": "elderConsoleLogErrors",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Write the html output to public.",
      "hook": "requestComplete",
      "name": "elderWriteHtmlFileToPublic",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Page generating timings and logging.",
      "hook": "requestComplete",
      "name": "elderDisplayRequestTime",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "A breakdown of the average times of different stages of the build.",
      "hook": "buildComplete",
      "name": "elderShowParsedBuildTimes",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Writes out any errors of a build to a JSON file in the ___ELDER___ folder.",
      "hook": "buildComplete",
      "name": "elderWriteBuildErrors",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "hooks.js",
        "type": "hooks.js",
      },
      "description": "just for testing",
      "hook": "bootstrap",
      "name": "test hook from file",
      "run": [Function],
    },
  ],
  "markBootstrapComplete": [Function],
  "query": Object {},
  "routes": Object {
    "route-a": Object {
      "$$meta": Object {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": Array [
        Object {
          "description": "test",
          "hook": "routeHook",
          "name": "route hook",
          "run": [MockFunction],
        },
      ],
    },
    "test-a": Object {
      "$$meta": Object {
        "addedBy": "elder-plugin-upload-s3",
        "type": "plugin",
      },
      "all": Array [],
      "data": [Function],
      "template": "fakepath/Test.svelte",
      "templateComponent": [Function],
    },
    "test-b": Object {
      "$$meta": Object {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": Array [],
    },
  },
  "runHook": [Function],
  "server": [Function],
  "serverLookupObject": Object {},
  "settings": Object {
    "$$internal": Object {
      "hashedComponents": Object {
        "File1": "entryFile1",
        "File2": "entryFile2",
      },
    },
    "build": false,
    "debug": Object {
      "automagic": true,
    },
    "hooks": Object {
      "disable": Array [
        "randomHook",
      ],
    },
    "locations": Object {
      "buildFolder": "./__ELDER__/",
      "srcFolder": "./src/",
      "svelte": Object {
        "clientComponents": "./public/dist/svelte/",
        "ssrComponents": "./___ELDER___/compiled/",
      },
    },
    "plugins": Object {
      "elder-plugin-upload-s3": Object {
        "dataBucket": "elderguide.com",
        "deployId": "11111111",
        "htmlBucket": "elderguide.com",
      },
    },
    "server": Object {
      "prefix": "/dev",
    },
    "typescript": true,
  },
  "shortcodes": Array [
    Object {
      "$$meta": Object {
        "addedBy": "elder",
        "type": "elder",
      },
      "run": [Function],
      "shortcode": "svelteComponent",
    },
  ],
}
`;

exports[`#Elder srcPlugin found 1`] = `
Elder {
  "allRequests": Array [
    Object {
      "permalink": "/dev/",
      "route": "test-a",
      "slug": "/test",
      "type": "server",
    },
  ],
  "bootstrapComplete": Promise {},
  "data": Object {},
  "errors": Array [],
  "helpers": Object {
    "inlineSvelteComponent": [Function],
    "permalinks": Object {
      "route-a": [Function],
      "test-a": [Function],
      "test-b": [Function],
    },
  },
  "hooks": Array [
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds external helpers to helpers object",
      "hook": "bootstrap",
      "name": "elderAddExternalHelpers",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Builds the shortcode parser, parses shortcodes from the html returned by the route's html and appends anything needed to the stacks.",
      "hook": "shortcodes",
      "name": "elderProcessShortcodes",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds <meta charset=\\"UTF-8\\" /> to the head.",
      "hook": "stacks",
      "name": "elderAddMetaCharsetToHead",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Adds <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\" /> to the head.",
      "hook": "stacks",
      "name": "elderAddMetaViewportToHead",
      "priority": 10,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Sets up the default polyfill for the intersection observer",
      "hook": "stacks",
      "name": "elderAddDefaultIntersectionObserver",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "AddsSystemJs to beforeHydrateStack also add preloading of systemjs to the headStack.",
      "hook": "stacks",
      "name": "elderAddSystemJs",
      "priority": 1,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Creates an HTML string out of the Svelte layout and stacks.",
      "hook": "compileHtml",
      "name": "elderCompileHtml",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Log any errors to the console.",
      "hook": "error",
      "name": "elderConsoleLogErrors",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Write the html output to public.",
      "hook": "requestComplete",
      "name": "elderWriteHtmlFileToPublic",
      "priority": 100,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Page generating timings and logging.",
      "hook": "requestComplete",
      "name": "elderDisplayRequestTime",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "A breakdown of the average times of different stages of the build.",
      "hook": "buildComplete",
      "name": "elderShowParsedBuildTimes",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder.js",
        "type": "internal",
      },
      "description": "Writes out any errors of a build to a JSON file in the ___ELDER___ folder.",
      "hook": "buildComplete",
      "name": "elderWriteBuildErrors",
      "priority": 50,
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder-plugin-upload-s3",
        "type": "plugin",
      },
      "description": "just for testing",
      "hook": "customizeHooks",
      "name": "test hook",
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder-plugin-upload-s3",
        "type": "plugin",
      },
      "description": "just for testing",
      "hook": "bootstrap",
      "name": "test hook 2",
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "elder-plugin-upload-s3",
        "type": "plugin",
      },
      "description": "just for testing",
      "hook": "bootstrap",
      "name": "test hook 3",
      "run": [Function],
    },
    Object {
      "$$meta": Object {
        "addedBy": "hooks.js",
        "type": "hooks.js",
      },
      "description": "just for testing",
      "hook": "bootstrap",
      "name": "test hook from file",
      "run": [Function],
    },
  ],
  "markBootstrapComplete": [Function],
  "query": Object {},
  "routes": Object {
    "route-a": Object {
      "$$meta": Object {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": Array [
        Object {
          "description": "test",
          "hook": "routeHook",
          "name": "route hook",
          "run": [MockFunction],
        },
      ],
    },
    "test-a": Object {
      "$$meta": Object {
        "addedBy": "elder-plugin-upload-s3",
        "type": "plugin",
      },
      "all": [Function],
      "data": [Function],
      "permalink": [Function],
      "template": "fakepath/Test.svelte",
      "templateComponent": [Function],
    },
    "test-b": Object {
      "$$meta": Object {
        "addedBy": "routejs",
        "type": "route",
      },
      "hooks": Array [],
    },
  },
  "runHook": [Function],
  "server": [Function],
  "serverLookupObject": Object {
    "/dev/": Object {
      "permalink": "/dev/",
      "route": "test-a",
      "slug": "/test",
      "type": "server",
    },
  },
  "settings": Object {
    "$$internal": Object {
      "hashedComponents": Object {
        "File1": "entryFile1",
        "File2": "entryFile2",
      },
    },
    "build": false,
    "debug": Object {
      "automagic": true,
    },
    "hooks": Object {
      "disable": Array [
        "randomHook",
      ],
    },
    "locations": Object {
      "buildFolder": "./__ELDER__/",
      "srcFolder": "./src/",
      "svelte": Object {
        "clientComponents": "./public/dist/svelte/",
        "ssrComponents": "./___ELDER___/compiled/",
      },
    },
    "plugins": Object {
      "elder-plugin-upload-s3": Object {
        "dataBucket": "elderguide.com",
        "deployId": "11111111",
        "htmlBucket": "elderguide.com",
      },
    },
    "server": Object {
      "prefix": "/dev",
    },
    "typescript": true,
  },
  "shortcodes": Array [
    Object {
      "$$meta": Object {
        "addedBy": "elder",
        "type": "elder",
      },
      "run": [Function],
      "shortcode": "svelteComponent",
    },
  ],
}
`;
